### PR TITLE
Add Token History tab

### PR DESCRIPTION
**Pull request by Andrew-128**

### CODE REVIEW

File: electroncash/wallet.py [hunk 1]
```diff
@@ -1047,6 +1053,16 @@ def _get_wallet_delta(self, tx, *, ver=1) -> Union[WalletDelta, WalletDelta2]:
                         value = v
                         if ver == 2:
                             spends_coins_mine.append(f'{prevout_hash}:{prevout_n}')
+                        if ver == 3:
+                            inputtx = self.get_input_tx(prevout_hash)
```
Comment (cculianu) [inline electroncash/wallet.py:1057]: `get_input_tx` may return `None`.. this should be handled somehow.

File: electroncash/wallet.py [hunk 2]
```diff
@@ -1021,19 +1021,25 @@ def get_tx_delta(self, tx_hash, address):
...
         """ Effect of tx on wallet """
-        assert ver in (1, 2)
+        assert ver in (1, 2, 3)
         is_relevant = False
         is_mine = False
         is_pruned = False
         is_partial = False
         v_in = v_out = v_out_mine = 0
         spends_coins_mine = list()
+        tokens_delta = defaultdict(lambda: {"fungibles": 0, "nfts_in": [], "nfts_out": []})  # key: category_id
```
Comment (cculianu) [inline electroncash/wallet.py:1041]: Nice.

File: electroncash_gui/qt/main_window.py [hunk 3]
```diff
@@ -231,6 +233,7 @@ def add_optional_tab(tabs, tab, icon, description, name, default=True):
         self.gui_object.update_available_signal.connect(self.on_update_available)  # shows/hides the update_available_button, emitted by update check mechanism when a new version is available
         self.gui_object.new_window_initialized_signal.connect(self.new_window_initialized)
         self.history_list.setFocus(True)
+        self.token_history_list.setFocus(True)
```
Comment (cculianu) [inline electroncash_gui/qt/main_window.py:236]: Probably you were just doing everything history_list did here.. and weren't aware of what `setFocus` does..

`setFocus` grabs keyboard focus .. considering that the history_list is up and not the token_history_list.. this line should be removed.

```suggestion
```

File: electroncash_gui/qt/main_window.py [hunk 4]
```diff
@@ -497,6 +501,7 @@ def load_wallet(self):
                 self._warn_if_invalid_testnet_wallet()
         self.watching_only_changed()
         self.history_updated_signal.emit() # inform things like address_dialog that there's a new history
+        #self.toggle_tab(self.token_history_tab)
```
Comment (cculianu) [inline electroncash_gui/qt/main_window.py:504]:
```suggestion
```

**Comment by cculianu** [comment]
Note -- there's also a method that you need to update: `ElectrumWindow._update_labels` in `main_window.py`.  It will then fire to trigger a refresh when labels change for the `token_history_list`.. e.g. you need to add a call `token_histroy_list.update()` in that `_update_labels` method.

**Comment by cculianu** [comment]
Memory usage caveat:  I got bitten by something along these lines before and so it's important to actually take a deep copy of a tx.

So `wallet.get_input_tx()` loads the tx from the wallet's `transactions` dictionary. The `Transaction` objects are stored there in a compact serialized form. Once you retrieve such an object, you get a shared reference to the same `Transaction` that lives in the wallet. So far so good -- however when one calls `tx.outputs()`, the `Transaction` class deserializes the object in-place and modifies the same instance that lives in the wallet's `transactions` dict.

What's the problem here?  Well -- I got bitten by this in the past. It ends up wasting *tons* of memory if you have many thousands or tens of thousands of txns.

If you iterate through all the transactions in the wallet (which may end up happening in, say, some heavy token-using wallet), and call `get_input_tx()` a lot you end up with a lot of memory bload.

I suggest adding a line somewhere in the implementation of `wallet.get_input_tx` to, in the case where it's an in-wallet instance (but not in the case where it came from the network) -- to basically do a `tx = copy.deepcopy(tx)`...

Let me know ...

File: electroncash_gui/qt/main_window.py (hunk 3 above)
Comment (Andrew-128) [reply electroncash_gui/qt/main_window.py:236]: Ah ok, and indeed, mostly this is based on history_list!

File: electroncash_gui/qt/main_window.py (hunk 4 above)
Comment (Andrew-128) [reply electroncash_gui/qt/main_window.py:503]: Oops! Thanks.

**Comment by cculianu** [comment]
Actually -- regarding the `get_input_tx` modification -- don't worry about it. I'll do it. The issue is that taking a deep copy each time and then deserializing for each iteration may eat CPU. So we save memory at the expense of CPU.

I'll rewrite this method a little bit to cache "recent" answers.

**Comment by cculianu** [comment]
OK, will merge now. I'll handle modifying `wallet.get_input_tx` to return a recently-cached copy if available (saves on network usage and also handles the memory bloat issue).

**Pull request merged**